README.md
# heirloom linuxxdoom v1.10 - The original open source DOOM

<!--
  Copyright (C) 2023 by Luke Shumaker <lukeshu@lukeshu.com>

  This program is free software; you can redistribute it and/or
  modify it under the terms of the GNU General Public License
  as published by the Free Software Foundation; either version 2
  of the License, or (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
-->

In 1994, id Software's Dave Taylor released two ports of DOOM v1.8 to
the then-new Linux kernel, which had just hit "1.0" earlier that year.
One port was `linuxsdoom`, which used SVGAlib; and the other was
`linuxxdoom`, which used X11.

In 1997, Bernd Kreimeier was given a 1997-01-10 snapshot of the DOOM
source code, which contained the code for the MS-DOS version,
linuxsdoom, and linuxxdoom (and maybe the NeXTSTEP version?).  That
snapshot does not precisely correspond to any released version of
DOOM.  He stripped out parts that id wasn't allowed to or declined to
share, stripped out the versions other than linuxxdoom, cleaned it up,
got it working again (bit-rot is a terrible thing).  Bernd Kreimeier's
modified linuxxdoom sources were then published by id on 1997-12-23,
with a README from id's John Carmack.

That 1997 Bernd-Kreimeier-modified source release is the basis for all
DOOM source ports (... until 2009 when Bethesda bought id, and
subsequently started releasing "official" but inferior ports based on
pre-open-source code).

heirloom linuxxdoom is a project to keep that original source release
from bit-rotting, while remaining as faithful as possible to that
original source release; out of historical interest.  It does not aim
to be faithful to any historical released binary; see [Chocolate
Doom][] if you are interested in that.

## Limitations

 - It does not support the mouse, it's keyboard-only (TODO: linuxxdoom
   v1.8 didn't support the mouse, but there's code there for the
   mouse.  Did it never work, or is this bit-rot?)

 - There is no music (but sound effects do work).  (Neither linuxxdoom
   nor linuxsdoom ever had music.)

 - It only supports OSS audio; rather than modern ALSA/PipeWire; you
   will need to enable OSS.

 - It only supports 8-bit PseudoColor X11; while modern versions of
   the Xorg server still support 8-bit PseudoColor, they can't enable
   it and modern 24-bit TrueColor at the same time.

 - It only supports x86-32 (TODO: Get it compiling for other
   architectures.  The change in `int` sizes is bit-rot.)

 - Demos only play if the WAD file version matches the executable
   version.  The latest official id WAD files, as well as most
   community WAD files, have version `109` (1.9), while linuxxdoom
   1.10 naturally has version `110`, so demos will generally not play.

 - It only supports rendering at 320x200, 640x400, or 960x600; it
   cannot go larger than that, and it cannot render at fractions.

 - DOOM's graphics are designed for non-square pixels (tall 5:6
   pixels); your modern display probably has square pixels; we make no
   attempt to stretch the image or otherwise compensate for this.

heirloom linuxxdoom has no intention to add music, move away from OSS,
move away from 8-bit PseudoColor X11, fuss with `VERSION`, et c.; See
the [TODO: link-here] project if you are interested in any of those
things.  Adding/fixing mouse support is dependent on why it's not
working.

## Compiling

If you are on x86-64, you will need do install the normal dev tools
(Make, GCC, ...), as well as the 32-bit versions of the libraries that
DOOM needs:

 - Arch Linux and derivatives: `sudo pacman -S --needed base-devel lib32-libxext lib43-libnsl`

If you are on a non-x86 platform, you won't be able to build this
without the help of an x86 emulator.

Once the dependencies are installed, simply run `make` in the
`linuxdoom-1.0/` and `sndserv/` directories:

    make -C linuxdoom-1.10/ && make -C sndserv/

The two executables that it builds are
`./linuxdoom-1.10/linux/linuxxdoom` and `./sndserv/linux/sndserver`.

## Installing

 - You will likely want to put the `linuxxdoom` executable somewhere
   in your `$PATH`.

 - I wish I had a good place to tell you to put the `sndserver`
   executable, but it needs to be in either the current-directory when
   you run `linuxxdoom`, or in the directory pointed to by your
   `$DOOMWADDIR` environment variable; adding it to your `$PATH` does
   NOT work.  Or, if you don't care about sound effects, you can just
   ignore the `sndserver`

 - Similarly, in either the current-directory or the `$DOOMWADDIR`
   directory, you will need a WAD file, with one of the following
   filenames:

   | filename       | what that filename usually is                  | notes                              |
   |----------------|------------------------------------------------|------------------------------------|
   | `doom1.wad`    | "DOOM" Shareware version (1993)                | just the 1st episode               |
   | `doom.wad`     | "DOOM" (1993)                                  | all 3 episodes                     |
   | `doomu.wad`    | "The Ultimate DOOM" (1995)                     | the 3 episodes, plus a 4th episode |
   | `plutonia.wad` | "Final DOOM: The Plutonia Experiment" (1996)   | a standalone episode               |
   | `tnt.wad`      | "Final DOOM: TNT: Evilution" (1996)            | a standalone episode               |
   | `doom2.wad`    | "DOOM II: Hell on Earth" (1994)                |                                    |
   | `doom2f.wad`   | "DOOM II: Hell on Earth" French version (1994) |                                    |

   Of course, any of those may be a community-created/modified WAD,
   rather than what it "usually" is.

 - Assuming that you don't want to turn off 24-bit color, you'll need
   a nested X server to run with 8-bit color, such as Xephyr:

   + Arch Linux and derivatives: `sudo pacman -S --needed xorg-server-xephyr`

## Running

 - You'll need to ensure that OSS audio is enabled:

        sudo modprobe snd_pcm_oss

 - You'll need to start that nested X server:

        Xephyr :1 -screen 640x400x8 &

   That `:1` should be set to an available display number; your main X
   display is probably `:0`, so `:1` is probably available.

   For the resolution, you can choose any of the supported
   resolutions: `320x200`, `640x400`, or `960x600`; but you *must*
   have the `x8` at the end to indicate 8-bit color.

 - Either `cd` to, or set `DOOMWADDIR` to the directory where you
   placed `sndserver` and the WAD file(s).

 - Set `DISPLAY` to whatever you chose for Xephyr above, and run the
   `linuxxdoom` executable; see [`./README.linuxx-1.8.txt`][] for the
   various options when running it.  By default it will run at
   320x200; to run at 640x400, give `-2` as an argument; to run at
   960x600, give `-3` as an argument.

        DISPLAY=:1 linuxxdoom -2

## The original documentation

Dave Taylor's Linux DOOM:

 - [`./README.linuxs-1.8.txt`][]: Dave Taylor's README for linuxsdoom
   v1.8.

 - [`./README.linuxx-1.8.txt`][]: Dave Taylor's README for linuxxdoom
   v1.8 (see below for a description of ways this file is
   out-of-date).

 - [`./README.dos-1.6.txt`][]: The README for DOS DOOM v1.6, which was
   included with linuxsdoom and linuxxdoom v1.8, and covers details
   that are not specific to Linux DOOM.


Bernd Kreimeier's source release:

 - [`./README.carmack.txt`][]: John Carmack's notes on the source release.

 - Bernd Kreimeier's notes on the state of the code, and the changes
   he made:

   + [`./linuxdoom-1.10/README.b`][]: His main README.
   + [`./linuxdoom-1.10/README.book`][]: The book he was writing, and
     how the source-release came to be.
   + [`./linuxdoom-1.10/README.sound`][]: Describes how sound works.
   + [`./linuxdoom-1.10/README.asm`][]: Some assembly routines from
     the DOS version of DOOM.
   + [`./linuxdoom-1.10/README.gl`][]: Notes on what would need to be
     done to port DOOM to OpenGL.
   + [`./sndserv/README.sndserv`][]: Basic introduction of the sound
     server.


## Differences from the linuxxdoom v1.8 README to v1.10

Surely the state of bugs and such between v1.8 and v1.10 has changed
(TODO: document them), but what follows are notes about things in
README.linuxx-1.8.txt that are not true for v1.10 (were they even true
for 1.8?).

 - The X server supporting the MIT-SHM extension is no longer
   required; it will be used if it does support it, but will
   gracefully fall back if the server doesn't support it.

 - The `sndserver` binary doesn't need to be in your path, instead it
   needs to be in either the current directory or the `$DOOMWADDIR`
   directory.

## License

In 1999, id Software authorized Andrew Stine to re-license the source
release to the GNU General Public License (GPL), and so that is the
license that heirloom linuxxdoom is under.

Unless the copyright statement in a specific file is edited, the
heirloom linuxxdoom project does not consider its patches to be
sufficiently creative to qualify for copyright, and are in the public
domain.  Feel free to cherry-pick them in to your own DOOM project
without the legal need for attribution (but attribution would be
appreciated).

[Chocolate Doom]: https://github.com/chocolate-doom/chocolate-doom
[`./README.linuxs-1.8.txt`]: ./README.linuxs-1.8.txt
[`./README.linuxx-1.8.txt`]: ./README.linuxx-1.8.txt
[`./README.dos-1.6.txt`]: ./README.dos-1.6.txt
[`./README.carmack.txt`]: ./README.carmack.txt
[`./linuxdoom-1.10/README.b`]: ./linuxdoom-1.10/README.b
[`./linuxdoom-1.10/README.book`]: ./linuxdoom-1.10/README.book
[`./linuxdoom-1.10/README.sound`]: ./linuxdoom-1.10/README.sound
[`./linuxdoom-1.10/README.asm`]: ./linuxdoom-1.10/README.asm
[`./linuxdoom-1.10/README.gl`]: ./linuxdoom-1.10/README.gl
[`./sndserv/README.sndserv`]: ./sndserv/README.sndserv


linuxdoom-1.10/ChangeLog



	* TODO: see below, and in the "TODO" file. Enjoy!

Mon Dec 22 20:29:16 1997    <bk@gamers.org>

	* CVS logs and other obsolete stuff removed. Anybody
 	  who wants to keep some revision control now has a
 	  clean slate to start with.

Mon Dec 22 19:53:34 1997    <bk@gamers.org>


	* i_sound.c: enabled SNDSERV, as SNDINTR for
 	  some reason just gives ghastly results e.g.
	  on E4M2. Frankly, I am at a loss. SNDSERV is
	  now default, until the internal sound driver
	  is a bit more reliable.
	  Note that the current redundancy means that
	  changes like the one below will have to
	  be propagated manually to the soundserver
	  sources.

	* m_menu.c: the 4th episode is now removed with
	  the original doom.wad. You need to rename the
	  Ultimate DOOM/Special Edition retail IWAD to
	  doomu.wad now, or you won't see the 4th episode
	  in the menu. The compile time SPECIAL define
	  is thus gone.

Mon Dec 22 17:08:33 1997    <bk@gamers.org>

	* v_video.c (V_DrawPatch): another last minute hack.
	  While shareware, retail, commercial, and plutonia
	  (being a full DOOM2 IWAD) seem to work okay now,
	  TNT gives an error on finishing the first mission:
	   "Patch at -35, -5 exceeds LFB".
	  I changed the error abort into a simple return,
	  thus the patch is ignored. The intermission screen
	  seems to come up okay.
	* TODO: check which patch, and whether it is an IWAD
	  problem.

	* i_sound.c: the sound table is hardwired in
	  sounds.h/sounds.c. As our current crude
 	  sound handling simply loads *all* sounds at
	  startup, we are going to miss some with DOOM1
	  WAD files. I could skip them, but decided to
	  load a placeholder instead (dspistol). It might
	  be good to use a distinct default sound for
	  WAD debug purposes. A zero length sound lump
	  would work, but would not be noticeable.
	  Anyway, shareware and retail work now.
	* TODO: implement proper handling for missing
	  lumps, sound and otherwise.
	  Perhaps move sound table into WAD?

	* g_game.c (G_DoPlayDemo): finally removed the
	  annoying "Demo is from a different game version"
	  abort. It now simply declines to playback the
	  demo, and waits for user input on some
	  do_nothing screen.

	* doomdef.h&Cie.: Lesson of the day - do not
	  replace a bunch of booleans with an enum and
	  use the same identifiers. Point in case:
	   "if ( commercial )" will not give an error,
	  and will always be true as long as the enum
	  value is greater than zero.
	  I found that the DOOM2 vs. DOOM differences
	  are everywhere (weapons, monsters, doors).
	  Number of episodes varies from shareware/commercial
	  to registered to retail, while commercial has
 	  a unique set (two of them, counting the german
	  edition) of maps in one episode. Plus, TNT and
	  Plutonia add some TITLE strings to the mixture.

	  Well, Plutonia and TNT are treated as DOOM2 for
	  now, so you will miss the startup message.

	* wi_stuff.h (NUMEPISODES): removed SPECIAL switch.
	  It is no 4 times 9 for wi_stuff.c internal
	  static arrays - doesn't matter.
	* TODO: unified handling with DOOM 2 - dynamic
	  allocation of arrays.

	* i_sound.c (I_UpdateSound): okay, I separated
 	  the mixing, now done synchonously, along with
	  a flag signalling the timer that the mixing buffer
	  has been updated. The handler is now very short,
	  and I tried several intervals down to 50usecs,
	  w/o complaints. Now the man page says:
	   "system timer resolution currently 10ms". Odd.
	  Anyway, while the double shotgun/plasma rapid
	  fire problem seems to be a bit less disturbing
	  at higher refresh, it's still there. I set the
	  interval to 500usec, which is sufficient for
	  avoiding any buffer update misses.
	  Conclusion after just two days of experimentation:
	  yep, sound driver code isn't fun at all.

	  As for the bug - well, Dave Taylor suggested
	  close distance getting into a divide-by-near-zero
	  situation, screwing up the volume. I can't figure
	  why latency of an external sound driver or screen
	  size affect this, but I am running out of ideas.

	* i_sound.c:
	  Some more experimentation with the timer driven
	  sound. It doesn't work well using an intervall
	  of less then 30 msecs - there will be artifacts
	  with say 50 msecs. This is pretty obvious with
 	  a target frame rate of at least 30fps, methinks.
	  Using the REAL/SIGALRM timer with 30msec gets
	  rid of the artifacts, it seems - at the expense
	  of slowing down things on a P133 to a noticeable
	  jerkiness. Bah.

Mon Dec 22 00:36:54 1997    <bk@gamers.org>

	* info.c: and i_video.c and i_sound.c - don't ask
	  me why some Linux header files are different with
	  gcc vs. g++, or what the complaint about the g++
  	  complaint info.c state table is all about:
	  "initializer element for `states[..].action.acp1'
	   is not constant"
	  Undid some changes, compiled with gcc, playtested,
	  seems okay. Done for today... yesterday.

	* i_net.c (ntohl): okay, htons/htonl, ntohs,ntohl
	  are back to haunt me. Copied the macros that
	  on my box aren't used for whatever reason directly
	  into the source. Got rid of all other multiple and
	  undefined references. CC=g++ now compiles (still
	  many warnings) and links, but the binary dumps a
	  core after Init PlayLoop. So be it.

Sun Dec 21 12:38:08 1997    <bk@gamers.org>

	* p_enemy.c (P_NewChaseDir): changed tdir to int,
	  removed the LUTs - spurious locks were due to
	  endless loops created by boneheaded predecessor
	  map. Has to be a better way to do enum dirtype_t
	  anyway. Problem seems to be fixed.

	* CC=gcc again, this time loads of #includes to
	  fix "implicit declarations, and one or two
	  unused variables. DOOM now compiles without
	  any -Wall warnings left, as C.

	* Bug: compiled the reworked code with gcc. Within a
	  solid while of testing and blasting away, it
	  locked once. Got a core, which gdb doesn't grok.
	  Bah.

	* TODO: okay, linkage of g++ build modules give loads

doomgeneric/sha1.c
/* sha1.c - SHA1 hash function
 *	Copyright (C) 1998, 1999, 2000, 2001 Free Software Foundation, Inc.
 *
 * Please see below for more legal information!
 *
 * This file is part of GnuPG.
 *
 * GnuPG is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * GnuPG is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, see <http://www.gnu.org/licenses/>.
 */


/*  Test vectors:
 *
 *  "abc"
 *  A999 3E36 4706 816A BA3E  2571 7850 C26C 9CD0 D89D
 *
 *  "abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq"
 *  8498 3E44 1C3B D26E BAAE  4AA1 F951 29E5 E546 70F1
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>

#include "i_swap.h"
#include "sha1.h"

void SHA1_Init(sha1_context_t *hd)
{
    hd->h0 = 0x67452301;
    hd->h1 = 0xefcdab89;
    hd->h2 = 0x98badcfe;
    hd->h3 = 0x10325476;
    hd->h4 = 0xc3d2e1f0;
    hd->nblocks = 0;
    hd->count = 0;
}


/****************
 * Transform the message X which consists of 16 32-bit-words
 */
static void Transform(sha1_context_t *hd, byte *data)
{
    uint32_t a,b,c,d,e,tm;
    uint32_t x[16];

    /* get values from the chaining vars */
    a = hd->h0;
    b = hd->h1;
    c = hd->h2;
    d = hd->h3;
    e = hd->h4;

#ifdef SYS_BIG_ENDIAN
    memcpy(x, data, 64);
#else
    {
        int i;
        byte *p2;
        for(i=0, p2=(byte*)x; i < 16; i++, p2 += 4 )
        {
            p2[3] = *data++;
            p2[2] = *data++;
            p2[1] = *data++;
            p2[0] = *data++;
        }
    }
#endif


#define K1  0x5A827999L
#define K2  0x6ED9EBA1L
#define K3  0x8F1BBCDCL
#define K4  0xCA62C1D6L
#define F1(x,y,z)   ( z ^ ( x & ( y ^ z ) ) )
#define F2(x,y,z)   ( x ^ y ^ z )
#define F3(x,y,z)   ( ( x & y ) | ( z & ( x | y ) ) )
#define F4(x,y,z)   ( x ^ y ^ z )

#define rol(x,n) ( ((x) << (n)) | ((x) >> (32-(n))) )

#define M(i) ( tm =   x[i&0x0f] ^ x[(i-14)&0x0f] \
               ^ x[(i-8)&0x0f] ^ x[(i-3)&0x0f] \
	       , (x[i&0x0f] = rol(tm,1)) )

#define R(a,b,c,d,e,f,k,m)  do { e += rol( a, 5 )     \
            + f( b, c, d )  \
            + k	      \
            + m;	      \
        b = rol( b, 30 );    \
    } while(0)
    R( a, b, c, d, e, F1, K1, x[ 0] );
    R( e, a, b, c, d, F1, K1, x[ 1] );
    R( d, e, a, b, c, F1, K1, x[ 2] );
    R( c, d, e, a, b, F1, K1, x[ 3] );
    R( b, c, d, e, a, F1, K1, x[ 4] );
    R( a, b, c, d, e, F1, K1, x[ 5] );
    R( e, a, b, c, d, F1, K1, x[ 6] );
    R( d, e, a, b, c, F1, K1, x[ 7] );
    R( c, d, e, a, b, F1, K1, x[ 8] );
    R( b, c, d, e, a, F1, K1, x[ 9] );
    R( a, b, c, d, e, F1, K1, x[10] );
    R( e, a, b, c, d, F1, K1, x[11] );
    R( d, e, a, b, c, F1, K1, x[12] );
    R( c, d, e, a, b, F1, K1, x[13] );
    R( b, c, d, e, a, F1, K1, x[14] );
    R( a, b, c, d, e, F1, K1, x[15] );
    R( e, a, b, c, d, F1, K1, M(16) );
    R( d, e, a, b, c, F1, K1, M(17) );
    R( c, d, e, a, b, F1, K1, M(18) );
    R( b, c, d, e, a, F1, K1, M(19) );
    R( a, b, c, d, e, F2, K2, M(20) );
    R( e, a, b, c, d, F2, K2, M(21) );
    R( d, e, a, b, c, F2, K2, M(22) );
    R( c, d, e, a, b, F2, K2, M(23) );
    R( b, c, d, e, a, F2, K2, M(24) );
    R( a, b, c, d, e, F2, K2, M(25) );
    R( e, a, b, c, d, F2, K2, M(26) );
    R( d, e, a, b, c, F2, K2, M(27) );
    R( c, d, e, a, b, F2, K2, M(28) );
    R( b, c, d, e, a, F2, K2, M(29) );
    R( a, b, c, d, e, F2, K2, M(30) );
    R( e, a, b, c, d, F2, K2, M(31) );
    R( d, e, a, b, c, F2, K2, M(32) );
    R( c, d, e, a, b, F2, K2, M(33) );
    R( b, c, d, e, a, F2, K2, M(34) );
    R( a, b, c, d, e, F2, K2, M(35) );
    R( e, a, b, c, d, F2, K2, M(36) );
    R( d, e, a, b, c, F2, K2, M(37) );
    R( c, d, e, a, b, F2, K2, M(38) );
    R( b, c, d, e, a, F2, K2, M(39) );
    R( a, b, c, d, e, F3, K3, M(40) );
    R( e, a, b, c, d, F3, K3, M(41) );
    R( d, e, a, b, c, F3, K3, M(42) );
    R( c, d, e, a, b, F3, K3, M(43) );
    R( b, c, d, e, a, F3, K3, M(44) );
    R( a, b, c, d, e, F3, K3, M(45) );
    R( e, a, b, c, d, F3, K3, M(46) );
    R( d, e, a, b, c, F3, K3, M(47) );
    R( c, d, e, a, b, F3, K3, M(48) );
    R( b, c, d, e, a, F3, K3, M(49) );
    R( a, b, c, d, e, F3, K3, M(50) );
    R( e, a, b, c, d, F3, K3, M(51) );
    R( d, e, a, b, c, F3, K3, M(52) );
    R( c, d, e, a, b, F3, K3, M(53) );
    R( b, c, d, e, a, F3, K3, M(54) );
    R( a, b, c, d, e, F3, K3, M(55) );
    R( e, a, b, c, d, F3, K3, M(56) );
    R( d, e, a, b, c, F3, K3, M(57) );
    R( c, d, e, a, b, F3, K3, M(58) );
    R( b, c, d, e, a, F3, K3, M(59) );
    R( a, b, c, d, e, F4, K4, M(60) );
    R( e, a, b, c, d, F4, K4, M(61) );
    R( d, e, a, b, c, F4, K4, M(62) );
    R( c, d, e, a, b, F4, K4, M(63) );
    R( b, c, d, e, a, F4, K4, M(64) );
    R( a, b, c, d, e, F4, K4, M(65) );
    R( e, a, b, c, d, F4, K4, M(66) );
    R( d, e, a, b, c, F4, K4, M(67) );
    R( c, d, e, a, b, F4, K4, M(68) );
    R( b, c, d, e, a, F4, K4, M(69) );
    R( a, b, c, d, e, F4, K4, M(70) );
    R( e, a, b, c, d, F4, K4, M(71) );
    R( d, e, a, b, c, F4, K4, M(72) );
    R( c, d, e, a, b, F4, K4, M(73) );
    R( b, c, d, e, a, F4, K4, M(74) );
    R( a, b, c, d, e, F4, K4, M(75) );
    R( e, a, b, c, d, F4, K4, M(76) );
    R( d, e, a, b, c, F4, K4, M(77) );
    R( c, d, e, a, b, F4, K4, M(78) );
    R( b, c, d, e, a, F4, K4, M(79) );

    /* update chainig vars */
    hd->h0 += a;
    hd->h1 += b;
    hd->h2 += c;
    hd->h3 += d;
    hd->h4 += e;
}


/* Update the message digest with the contents
 * of INBUF with length INLEN.
 */
void SHA1_Update(sha1_context_t *hd, byte *inbuf, size_t inlen)
{
    if (hd->count == 64)
    {
        /* flush the buffer */
	Transform(hd, hd->buf);
	hd->count = 0;
	hd->nblocks++;
    }
    if (!inbuf)
	return;
    if (hd->count)
    {
	for (; inlen && hd->count < 64; inlen--)
	    hd->buf[hd->count++] = *inbuf++;
	SHA1_Update(hd, NULL, 0);
	if (!inlen)
	    return;
    }

    while (inlen >= 64)
    {
	Transform(hd, inbuf);
	hd->count = 0;
	hd->nblocks++;
	inlen -= 64;
	inbuf += 64;
    }
    for (; inlen && hd->count < 64; inlen--)
	hd->buf[hd->count++] = *inbuf++;
}


/* The routine final terminates the computation and
 * returns the digest.
 * The handle is prepared for a new cycle, but adding bytes to the
 * handle will the destroy the returned buffer.
 * Returns: 20 bytes representing the digest.
 */

void SHA1_Final(sha1_digest_t digest, sha1_context_t *hd)
{
    uint32_t t, msb, lsb;
    byte *p;

    SHA1_Update(hd, NULL, 0); /* flush */;

    t = hd->nblocks;
    /* multiply by 64 to make a byte count */
    lsb = t << 6;
    msb = t >> 26;
    /* add the count */
    t = lsb;
    if ((lsb += hd->count) < t)
	msb++;
    /* multiply by 8 to make a bit count */
    t = lsb;
    lsb <<= 3;
    msb <<= 3;
    msb |= t >> 29;

    if (hd->count < 56)
    {
        /* enough room */
	hd->buf[hd->count++] = 0x80; /* pad */
	while (hd->count < 56)
	    hd->buf[hd->count++] = 0;  /* pad */
    }
    else
    {
        /* need one extra block */
	hd->buf[hd->count++] = 0x80; /* pad character */
	while (hd->count < 64)
	    hd->buf[hd->count++] = 0;
	SHA1_Update(hd, NULL, 0);  /* flush */;
	memset(hd->buf, 0, 56 ); /* fill next block with zeroes */
    }
    /* append the 64 bit count */
    hd->buf[56] = msb >> 24;
    hd->buf[57] = msb >> 16;
    hd->buf[58] = msb >>  8;
    hd->buf[59] = msb	   ;
    hd->buf[60] = lsb >> 24;
    hd->buf[61] = lsb >> 16;
    hd->buf[62] = lsb >>  8;
    hd->buf[63] = lsb	   ;
    Transform(hd, hd->buf);

    p = hd->buf;
#ifdef SYS_BIG_ENDIAN
#define X(a) do { *(uint32_t*)p = hd->h##a ; p += 4; } while(0)
#else /* little endian */
#define X(a) do { *p++ = hd->h##a >> 24; *p++ = hd->h##a >> 16;	 \
        *p++ = hd->h##a >> 8; *p++ = hd->h##a; } while(0)
#endif
    X(0);
    X(1);
    X(2);
    X(3);
    X(4);
#undef X

    memcpy(digest, hd->buf, sizeof(sha1_digest_t));
}

void SHA1_UpdateInt32(sha1_context_t *context, unsigned int val)
{
    byte buf[4];

    buf[0] = (val >> 24) & 0xff;
    buf[1] = (val >> 16) & 0xff;
    buf[2] = (val >> 8) & 0xff;
    buf[3] = val & 0xff;

    SHA1_Update(context, buf, 4);
}

void SHA1_UpdateString(sha1_context_t *context, char *str)
{
    SHA1_Update(context, (byte *) str, strlen(str) + 1);
}
